text/0000-let-expression.md

```diff
@@ -102,6 +102,46 @@ assert!(matches!(a, Some(x) if let Some(y) = b(x) && x == y));
 assert!(if let Some(x) = a && let Some(y) = b(x) && x == y { true } else { false });
 ```
 
+### Practical usage of this features
+
+People find let expressions theoretical at first look, and only traditional cases (if-let-chain and let-else)
+Can become useful. In this section there are some usages for new features of this RFC in real codes.
+
+This is an example from rust-clippy repository:
+```rust
+for w in block.stmts.windows(2) {
+    if_chain! {
+        if let StmtKind::Semi(first) = w[0].kind;
+        if let StmtKind::Semi(second) = w[1].kind;
+        if !differing_macro_contexts(first.span, second.span);
+        if let ExprKind::Assign(lhs0, rhs0, _) = first.kind;
+        if let ExprKind::Assign(lhs1, rhs1, _) = second.kind;
+        if eq_expr_value(cx, lhs0, rhs1);
+        if eq_expr_value(cx, lhs1, rhs0);
+        then {
+            // 30 lines of code with massive rightward drift
+        }
+    }
+}
+```
+
+Which by a generalized let-else can become:
+```rust
+for w in block.stmts.windows(2) {
+    (let StmtKind::Semi(first) = w[0].kind)
+    && (let StmtKind::Semi(second) = w[1].kind)
+    && !differing_macro_contexts(first.span, second.span)
+    && (let ExprKind::Assign(lhs0, rhs0, _) = first.kind)
+    && (let ExprKind::Assign(lhs1, rhs1, _) = second.kind)
+    && eq_expr_value(cx, lhs0, rhs1)
+    && eq_expr_value(cx, lhs1, rhs0)
+    || continue;
+    // 30 lines of code with two less tab
```

Comment:
this is more like a practical example of the *existing* `if_let_chain` feature

```rust
    if let StmtKind::Semi(first) = w[0].kind 
        && let StmtKind::Semi(second) = w[1].kind
        && !differing_macro_contexts(first.span, second.span)
        ...
        && eq_expr_value(cx, lhs1, rhs0)
    {
        continue;
    }
```

this is also more consistent with the way people would write

```rust
if next_condition {
    continue;
}
```

rather than

```
next_condition || continue;
```

Reply:
if-let-chain still need one indentation, because variables are bound inside of if body (like if-let) but in the example with `||` variables are bound after statement (like let-else). If you believe if-let can't meet the needs that let-else provide, The same statement is true for if-let-chain and this RFC.

> this is also more consistent with the way people would write

I am agree and prefer explicit `if` over `||` in normal cases. Without bindings, both behavior is the same (both in current rust and in this RFC) but because binding rules for them are different (for good reasons; for example accessing if-let variables after body is surprising) you can save one indentation by `||` one.

And it is worth noting that `!next_condition || continue;` is completely valid and working in today rust. But `!next_condition else { continue };` isn't and won't, so let-else syntax has a negative point here and this RFC is more consistent with rust.